### PR TITLE
Fix purge-request-specs manage cmd for multi-cell setups

### PR DESCRIPTION
**Pull request by grandchild**
Collect active-instance ids from other cells (other than
cell0 and cell1) as well. These ids are passed to the purge
function to leave the respective request_specs and, more
importantly, instance_mappings alone when cleaning up.

Failure to do so will delete instance_mappings for instances
in these other cells.

### CODE REVIEW

File: nova/cmd/manage.py [hunk 1]
```diff
@@ -831,8 +832,19 @@ def purge_request_specs(self, dry_run=False, older_than=90, max_number=0,
...
-        extant_instance_uuids = [i.get('uuid') for i in db.instance_get_all(
-            admin_context, columns_to_join=[]) if i.get('uuid') is not None]
+        extant_instance_uuids = set(i.get('uuid') for i in db.instance_get_all(
+            admin_context, columns_to_join=[]) if i.get('uuid') is not None)
+        # Fetch instances from other cells than the default cell1 and cell0
+        extra_cells = [cell for cell
+                       in cell_mapping_obj.CellMappingList.get_all(
+                            admin_context)
+                       if cell.name not in ["cell0", "cell1"]]
+        for cell in extra_cells:
```
Comment (joker-official) [inline nova/cmd/manage.py:842]: Why filter some cells out (hardcoded) here instead of looping over all cells to begin with?

Comment (grandchild) [reply nova/cmd/manage.py]: True. Fixed.

File: nova/cmd/manage.py [hunk 2]
```diff
@@ -829,10 +830,15 @@ def purge_request_specs(self, dry_run=False, older_than=90, max_number=0,
...
-        extant_instance_uuids = [i.get('uuid') for i in db.instance_get_all(
-            admin_context, columns_to_join=[]) if i.get('uuid') is not None]
+        for cell in cell_mapping_obj.CellMappingList.get_all(admin_context):
+            with context.target_cell(admin_context, cell) as cell_context:
+                # There's no need to join the default security_groups and
+                # instance_info_caches tables, hence explicit empty
+                # 'columns_to_join'.
+                extant_instance_uuids |= \
+                    set(i.get('uuid') for i in db.instance_get_all(
+                        cell_context, columns_to_join=[])
```
Comment (joker-official) [inline nova/cmd/manage.py:840]: Should raise `NameError` for undefined `extant_instance_uuids`.

Comment (grandchild) [reply nova/cmd/manage.py:841]: ... That's what I get for moving fast. :/

**Review by joker-official — COMMENTED** [review]
If this fails to connect to a cell, it will fail the command, right? Just asking, because I don't know the inner workings of the `context.target_cell()` context manager and now behaved badly for us before, when it ignored errors in cells..

**Comment by grandchild** [comment]
I am cautiously unsure.

The cell context _might_ be initialized wrongly if the transport to the cell is not available, but not throw an error. `oslo.messaging` does seem to simply prepare the connection objects and not actually call much into the DB or messaging backends. I might have overlooked things.

The cell context then contains a configured DB connection, which should AFAICS fail the call to `db.instance_get_all()` -- with an exception. I don't really know how to test this without a cell setup where I can bring down a DB or messaging pod.

**Review by joker-official — APPROVED** [review]
It doesn't seem to catch any exceptions, so I'm fine with this.

**Pull request merged**